docs/design.md

```diff
@@ -0,0 +1,196 @@
+
+Users can invoke the output command to stream the output of a process in real time (a combination
+of stdout and stderr). It gets the output right from the start. The command terminates once there
+is no more output to read i.e the process is done or has been terminated.
+
+```
+> job output 9de95676-84ed-4e37-b44c-4f4d1df7c74f
+test command output line 1
+test command output line 2
+test command output line 3
+test command output line 3
+…
+```
+
+## Resource Control
+
+We will use [cgroupsv2](https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html) to assign system resources to processes.
+
+It is a requirement that the host runs Linux kernel 4.15 (or above) as that is the version when the cpu controller was
+introduced.
+
+On startup, the worker will create a worker root cgroup which will optionally contain some 
+default values that can be provided via flags. For every process, we will create a new children of the
+worker cgroup with the provided resource control values. 
+
+We will enable users to control memory, IO and CPU. We will achieve this by exposing
+API parameters that allow assigning values to the following cgroup interface files:
+
+#### memory.max
+
+https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html#memory
+
+This is a memory usage hard limit. 
+
+#### io.max
+
+https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html#io
+
+We will allow users to set values for the rbps (Max read bytes per second) and wbps (Max write bytes per second) map keys in the io.max interface file.
+
+To allow users to set this option, we will need to start the worker with the $MAJ:$MIN device 
+number that we would like this limit to apply to.
+
+#### cpu.max
+
+https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html#cpu-interface-files
+
+This enables the distribution of CPU time by adjusting CPU bandwidth.
+
+The file contains 2 value which represent the allowed time quota (microseconds) and the 
+period length (microseconds).
+
+## GRPC API
+
+The service exposes a gRPC API that gives capabilities to start a process, stop a process, get process status and stream process output.
+
+The proposed proto file can be found here: [worker_official.proto](./worker.proto)
+
+The GRPC server will be calling methods in the worker library to interact with processes.
+
+### Authentication
+
+The service will use mTLS to secure traffic and verify client and server authenticity. For this design document, we will make the assumption that the Certificate Authority (CA) can be trusted i.e. certificates are only issued to users and nodes that have been verified.
+
+Both server and client will support TLS versions 1.2 and 1.3.
+
+The [grpc-go](https://github.com/grpc/grpc-go) package that we will be using to implement the server and client
+give direct access to the standard library's `tls.Config` struct so it should be relatively simple to verify both
+client and server certificates.
+
+### Cipher Suites
+
+#### TLS 1.2
```

Comment:
For the purposes of this exercise you might be better off ditching 1.2 support, rather than implementing a custom negotiation routine that filters out all of the undesirables. Although you could leave 1.2 support in as a stretch goal if you like.

Reply:
I think I will most likely not have enough time to implement stretch goals so I'd be happy to ditch 1.2 support for this exercise :)

Are you happy with that @codingllama?

Reply:
Sounds good to me.